app/gui2/src/components/GraphEditor/dragging.ts

```diff
@@ -0,0 +1,168 @@
+ * boundaries (to make node aligning easier for the user).
+ *
+ * The set of dragged nodes is determined upon construction. The new position is stored in the
+ * node metadata only when dragging is finished.
+ *
+ * The `offset` is always based of the nodes' initial positions.
+ */
+export function useDragging() {
+  const graphStore = useGraphStore()
+  const selection = injectGraphSelection(true)
+
+  // Logically, those fields could be inside CurrentDrag, but animations need component scope.
+  const offset = ref(Vec2.Zero)
+  const snapXTarget = ref(0)
+  const snapX = useApproach(snapXTarget, 30)
+  const snapYTarget = ref(0)
+  const snapY = useApproach(snapYTarget, 30)
+  const snappedOffset = computed(() => offset.value.add(new Vec2(snapX.value, snapY.value)))
+  const snappedOffsetTarget = computed(() =>
+    offset.value.add(new Vec2(snapXTarget.value, snapYTarget.value)),
+  )
+  class CurrentDrag {
+    draggedNodes: ExprId[]
+    grid: SnapGrid
+    stopPositionUpdate: WatchStopHandle
+
+    constructor(movedId: ExprId) {
+      markRaw(this)
+      this.draggedNodes = selection?.isSelected(movedId)
+        ? Array.from(selection.selected)
+        : [movedId]
+      this.grid = this.createSnapGrid()
+      offset.value = Vec2.Zero
+      snapXTarget.value = 0
+      snapYTarget.value = 0
+      snapX.skip()
+      snapY.skip()
+
+      this.stopPositionUpdate = watchEffect(() => {
+        for (const id of this.draggedNodes) {
+          const node = graphStore.nodes.get(id)
+          if (node == null) continue
+          node.visiblePosition = node.position.add(snappedOffset.value)
+        }
+      })
+    }
+
+    updateOffset(newOffset: Vec2): void {
+      const oldSnappedOffset = snappedOffset.value
+      const rects: Rect[] = []
+      for (const id of this.draggedNodes) {
+        const rect = graphStore.nodeRects.get(id)
+        const node = graphStore.nodes.get(id)
+        if (rect != null && node != null)
+          rects.push(new Rect(node.position.add(newOffset), rect.size))
+      }
+      const snap = this.grid.snappedMany(rects, DRAG_SNAP_THRESHOLD)
+      offset.value = newOffset
+      snapXTarget.value = snap.x
+      snapYTarget.value = snap.y
+      const newSnappedOffsetTarget = snappedOffsetTarget.value
+      // Skip animation if target offset does not change significantly, to avoid shivering
+      // when node is snapped.
+      if (abs(newSnappedOffsetTarget.x - oldSnappedOffset.x) < 2.0) {
+        snapX.skip()
+      }
+      if (abs(newSnappedOffsetTarget.y - oldSnappedOffset.y) < 2.0) {
+        snapY.skip()
+      }
+    }
+
+    finishDragging(): void {
+      this.stopPositionUpdate()
+      for (const id of this.draggedNodes) {
+        const node = graphStore.nodes.get(id)
+        if (node == null || node.visiblePosition == null) continue
+        const newPosition = node.position.add(snappedOffsetTarget.value)
+        graphStore.setNodePosition(id, newPosition)
```

Comment:
Real node position should also be updated during the drag, not only after it has been completed. Otherwise we lose the ability to synchronize that position with other clients. Additionaly, to prevent overwriting changes made by other clients during our drag (e.g. multiple users dragging, or a file change), the nodes for which the current position doesn't match the last position set during the drag should be excluded from that drag interaction.